lasy/utils/mode_decomposition.py

```diff
@@ -0,0 +1,133 @@
+from lasy.profiles.transverse.transverse_profile import TransverseProfile
+from lasy.profiles.transverse.transverse_profile_from_data import (
+    TransverseProfileFromData,
+)
+from lasy.profiles.transverse.hermite_gaussian_profile import (
+    HermiteGaussianTransverseProfile,
+)
+from lasy.utils.exp_data_utils import find_d4sigma
+
+import numpy as np
+import math
+
+
+def hermite_gauss_decomposition(laserProfile, n_x_max=12, n_y_max=12):
+    """
+    Decomposes a `lasy` laser profile into a set of hermite-gaussian
+    modes.
+
+    The function takes either an instance of `TransverseProfile` or an
+    instance of `Laser` (that is, either a transverse profile or the
+    full 3D laser profile defined on a grid). In the case that an
+    instance of `Laser` is passed then the intensity of this profile
+    is projected onto an x-y plane for the decomposition.
+
+    Parameters
+    ----------
+    laserProfile: class instance
+        An instance of a class or sub-class of TransverseLaserProfile
+
+    n_x_max, n_y_max: ints
+        The maximum values of `n_x` and `n_y` out to which the expansion
+        will be performed
+
+    Returns
+    -------
+    weights: dict of floats
+        A dictionary of floats corresponding to the weights of each mode
+        in the decomposition. The keys of the dictionary are tuples
+        corresponding to (`n_x`,`n_y`)
+
+    """
+
+    # Check if the provided laserProfile is a full laser profile or a
+    # transverse profile.
+
+    assert isinstance(
+        laserProfile, TransverseProfile
+    ), "laserProfile must be an instance of TransverseProfile"
+
+    # Here we need to define a grid size to use for the calculation
+    # By default we will use N x N grid where N = 500
+    N = 500
+
```

Comment:
Could this be an optional input parameter? Also, this should be mostly used mainly for experimentally measured profiles, right? When that's the case, should we store the original resolution in TransverseProfileFromData, and expose it here, to simplify the use and also avoid additional smoothing?

Reply:
I think the point on oversmoothing is a good one, although I think we should leave it to the user to decide the value of N. I will expose as an input parameter with a default of 500 and leave it to the user